xwiki-rendering-transformations/xwiki-rendering-transformation-icon/src/main/java/org/xwiki/rendering/internal/transformation/icon/IconTransformation.java

```diff
@@ -229,13 +235,14 @@ private void parseTree(List<Block> sourceBlocks)
                     }
                     count++;
                     mappingCursor = mappingCursor.getChildren().get(0);
-                    // If we reach the Image Block then we've found a match!
-                    if (mappingCursor instanceof ImageBlock) {
-                        // Replace the first source block with the image block and remove all other blocks...
+                    // If we reach the Icon Block then we've found a match!
+                    if (mappingCursor instanceof IconBlock) {
+                        // Replace the first source block with the icon block and remove all other blocks...
                         for (int i = 0; i < count - 1; i++) {
                             matchStartBlock.getParent().removeBlock(matchStartBlock.getNextSibling());
                         }
-                        sourceBlock = mappingCursor.clone();
+                        sourceBlock = mappingCursor
+                            .getFirstBlock(AnyBlockMatcher.ANYBLOCKMATCHER, Block.Axes.DESCENDANT).clone();
```

Comment:
Why are you cloning the first child of the icon block? What if the icon block has several children?

Reply:
The way I defined and used the IconBlock so far is that it contains only one child which is a block containing the whole icon.
---
After thinking about it, there's no reason to exclude other use cases where multiple blocks are given as children, possibly imposing an extra wrapping that's not needed.
---
After looking at the way we parse trees, it's not straigthforward to allow multiple children.
---
I have trouble understanding what this function does, there are a lot of abstract concepts and minimal documentation (and the recursive nature of the function does not make things simpler). With a simple test (1. make sure the default renderer behaves correctly, and make sure a renderer that add a random wordBlock behaves as expected) I couldn't find out an easy change that would return the expected value for tests on both cases...
I think it's okay to assume that all the content in the Icon block is nested inside a singular block.
I'm adding a line in the definition of the IconBlock to highlight this.

Reply:
We could just merge those children together in a composite block :thinking: looking into it...

Reply:
Okay, after taking a bit of distance, I figured out I could just wrap all the children block in a compositeBlock if needed. This works properly in both of the situations mentionned above :)

Reply:
Wrapping the children blocks in a composite block when they are multiple does not work either.

Reply:
Why don't you just clone the `IconBlock` itself, i.e., just revert the change on this line?

Reply:
Somehow it fails at transforming multiple icons one after the other:
https://up1.xwikisas.com/#6lC3HfY35P2Qv2BVGf6LKw